File: Backend/src/utils/crypto.ts
// src/utils/crypto.ts
import crypto from 'crypto'

/**
 * AES-256-GCM encryption for API keys at rest.
 *
 * Key source: ENCRYPTION_KEY env var (any length) → SHA-256 hashed to a
 * 32-byte key. When the var is absent we fall back to a fixed dev secret so
 * local development keeps working; deployments MUST set ENCRYPTION_KEY.
 *
 * Ciphertext format (base64, ':'-separated so JSON/keys stay safe):
 *   enc:v1:<iv>:<authTag>:<ciphertext>
 */
const ENCRYPTION_SECRET =
  process.env.ENCRYPTION_KEY

const KEY = crypto.createHash('sha256').update(ENCRYPTION_SECRET!).digest()

export const ENCRYPTED_PREFIX = 'enc:v1:'

/** Encrypt a plaintext secret. Returns `enc:v1:...` (or the input unchanged if already encrypted). */
export function encryptSecret(plain?: string): string {
  if (!plain) return ''
  if (plain.startsWith(ENCRYPTED_PREFIX)) return plain
  const iv = crypto.randomBytes(12)
  const cipher = crypto.createCipheriv('aes-256-gcm', KEY, iv)
  const ct = Buffer.concat([cipher.update(plain, 'utf8'), cipher.final()])
  const tag = cipher.getAuthTag()
  return `${ENCRYPTED_PREFIX}${iv.toString('base64')}:${tag.toString('base64')}:${ct.toString('base64')}`
}

/**
 * Decrypt an `enc:v1:` value.
 * - Returns null when the value is encrypted but cannot be decrypted (bad key / tampered).
 * - Returns the value unchanged when it is NOT in the encrypted format
 *   (legacy plaintext rows keep working until they are re-saved through the UI).
 */
export function decryptSecret(value?: string): string | null {
  if (!value) return null
  if (!value.startsWith(ENCRYPTED_PREFIX)) return value
  try {
    const [, ivB64, tagB64, ctB64] = value.split(':')
    const decipher = crypto.createDecipheriv('aes-256-gcm', KEY, Buffer.from(ivB64, 'base64'))
    decipher.setAuthTag(Buffer.from(tagB64, 'base64'))
    return Buffer.concat([
      decipher.update(Buffer.from(ctB64, 'base64')),
      decipher.final(),
    ]).toString('utf8')
  } catch {
    return null
  }
}

File: AiCalls/lib/crypto.py
# lib/crypto.py
"""AES-256-GCM encryption helpers shared with the Node backend.

Mirrors `Backend/src/utils/crypto.ts`:
  - key derived (sha256) from the ENCRYPTION_KEY env (falls back to a fixed dev
    secret so local dev works without extra config)
  - ciphertext format: enc:v1:<iv_b64>:<tag_b64>:<ct_b64>  (base64, ':' delimited)

The encryption key is derived lazily on first use — not at import time — so
that config.py's load_dotenv() (which runs during app boot) has already
populated ENCRYPTION_KEY by the time any encrypt/decrypt actually happens.
"""

import os
import hashlib
import base64
import threading

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

_PREFIX = "enc:v1:"
_LOCK = threading.Lock()
_KEY_CACHE: bytes | None = None


def _derive_key() -> bytes:
    global _KEY_CACHE
    if _KEY_CACHE is None:
        with _LOCK:
            if _KEY_CACHE is None:
                secret = os.getenv("ENCRYPTION_KEY")
                _KEY_CACHE = hashlib.sha256(secret.encode("utf-8")).digest()
    return _KEY_CACHE


def encrypt(plain: str) -> str:
    if not plain:
        return ""
    if plain.startswith(_PREFIX):
        return plain
    iv = os.urandom(12)
    # AESGCM.encrypt returns ciphertext with the 16-byte auth tag APPENDED
    # (ciphertext + tag). We store them separately to match the Node backend.
    combined = AESGCM(_derive_key()).encrypt(iv, plain.encode("utf-8"), None)
    ct = combined[:-16]
    tag = combined[-16:]
    return (
        f"{_PREFIX}"
        f"{base64.b64encode(iv).decode()}:"
        f"{base64.b64encode(tag).decode()}:"
        f"{base64.b64encode(ct).decode()}"
    )


def decrypt(value: str | None) -> str | None:
    """Decrypt an enc:v1: value.

    Returns None if it cannot be decrypted. Values NOT in the encrypted format
    are returned unchanged (legacy plaintext rows still work).
    """
    if not value:
        return None
    if not value.startswith(_PREFIX):
        return value
    try:
        rest = value[len(_PREFIX):]
        iv_b64, tag_b64, ct_b64 = rest.split(":")
        iv = base64.b64decode(iv_b64)
        tag = base64.b64decode(tag_b64)
        ct = base64.b64decode(ct_b64)
        # AESGCM.decrypt expects tag appended: ciphertext + tag
        combined = ct + tag
        plain = AESGCM(_derive_key()).decrypt(iv, combined, None)
        return plain.decode("utf-8")
    except Exception:
        return None
